File: Yixuan/19_01_17/recover_rotated_sorted_array.cpp
class Solution {
    /**
     * @param nums: The rotated sorted array
     * @return: void
     * @link: http://www.lintcode.com/en/problem/recover-rotated-sorted-array/
     * @author: Egbert Li
     */
public:
   void recoverRotatedSortedArray(vector<int> &nums) {
      int length = nums.size();
      for (int index = 0; index < length - 1; index++) {
         if (nums[index] > nums[index + 1]) {
            reverse(nums, 0, index);
            reverse(nums, index + 1, length - 1);
            reverse(nums, 0, length - 1);
         }
      }
   }

private:
   void reverse(vector<int> &nums, int start, int end) {
      while (start < end) {
         int temp = nums[start];
         nums[start] = nums[end];
         nums[end] = temp;
         start ++;
         end --;
      }
   }
};

File: Yixuan/19_01_17/recover_rotated_sorted_array.java
public class Solution {
    /**
     * @param nums: The rotated sorted array
     * @return: void
     * @link: http://www.lintcode.com/en/problem/recover-rotated-sorted-array/
     * @author: Egbert Li
     */
   public void recoverRotatedSortedArray(ArrayList<Integer> nums) {
      for (int index = 0; index < nums.size() - 1; index++) {
         if (nums.get(index) > nums.get(index + 1)) {
            reverse(nums, 0, index);
            reverse(nums, index + 1, nums.size() - 1);
            reverse(nums, 0, nums.size() - 1);
         }
      }
   }

   private void reverse(ArrayList<Integer> nums, int start, int end) {
      while (start < end) {
         int temp = nums.get(start);
         nums.set(start, nums.get(end));
         nums.set(end, temp);
         start ++;
         end --;
      }
   }
}
